File: fortran/sparse_memory.f90
!> Sparse memory layer inspired by UltraMem (ICLR 2025, ByteDance Seed team).
module sparse_memory
  implicit none
  integer, parameter, private :: dp = selected_real_kind(15, 307)

  type :: MemoryLayer
    real(dp), allocatable :: values(:,:,:)     ! (n_rows, n_cols, hidden_dim)
    real(dp), allocatable :: row_keys(:,:)     ! (n_ranks, hidden_dim) - projection
    real(dp), allocatable :: row_embed(:,:)    ! (n_ranks, n_rows) - row embeddings
    real(dp), allocatable :: col_keys(:,:)     ! (n_ranks, hidden_dim) - projection
    real(dp), allocatable :: col_embed(:,:)    ! (n_ranks, n_cols) - col embeddings
    real(dp), allocatable :: tucker_core(:,:)  ! (n_ranks, n_ranks)
    real(dp), allocatable :: linear_w(:,:)     ! (hidden_dim, 4) for IVE
    integer :: n_rows, n_cols, hidden_dim, n_ranks, top_k
  end type

contains

  subroutine init_memory(layer, n_rows, n_cols, hidden_dim, n_ranks, top_k)
    type(MemoryLayer), intent(inout) :: layer
    integer, intent(in) :: n_rows, n_cols, hidden_dim, n_ranks, top_k

    real(dp) :: scale

    layer%n_rows = n_rows
    layer%n_cols = n_cols
    layer%hidden_dim = hidden_dim
    layer%n_ranks = n_ranks
    layer%top_k = top_k

    allocate(layer%values(n_rows, n_cols, hidden_dim))
    allocate(layer%row_keys(n_ranks, hidden_dim))
    allocate(layer%row_embed(n_ranks, n_rows))
    allocate(layer%col_keys(n_ranks, hidden_dim))
    allocate(layer%col_embed(n_ranks, n_cols))
    allocate(layer%tucker_core(n_ranks, n_ranks))
    allocate(layer%linear_w(hidden_dim, 4))

    scale = sqrt(2.0_dp / real(n_ranks + hidden_dim, dp))

    call random_number(layer%values)
    layer%values = (layer%values - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%row_keys)
    layer%row_keys = (layer%row_keys - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%row_embed)
    layer%row_embed = (layer%row_embed - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%col_keys)
    layer%col_keys = (layer%col_keys - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%col_embed)
    layer%col_embed = (layer%col_embed - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%tucker_core)
    layer%tucker_core = (layer%tucker_core - 0.5_dp) * 2.0_dp * scale

    call random_number(layer%linear_w)
    layer%linear_w = (layer%linear_w - 0.5_dp) * 2.0_dp * scale

  end subroutine init_memory

  subroutine query(layer, query_vec, output_vec, n_query)
    type(MemoryLayer), intent(in) :: layer
    integer, intent(in) :: n_query
    real(dp), intent(in)  :: query_vec(layer%hidden_dim, n_query)
    real(dp), intent(out) :: output_vec(layer%hidden_dim, n_query)

    real(dp), allocatable :: scores(:,:)
    real(dp), allocatable :: top_scores(:)
    integer, allocatable  :: top_indices(:)
    integer :: q, k, r, c, idx
    real(dp) :: total_score

    allocate(scores(layer%n_rows, layer%n_cols))
    allocate(top_indices(layer%top_k))
    allocate(top_scores(layer%top_k))

    output_vec = 0.0_dp

    do q = 1, n_query
      ! Compute full score matrix via TDQKR
      call compute_all_scores(layer, query_vec(:,q), scores)

      ! Find top-k (flatten 2D -> 1D indices)
      call find_top_k_2d(scores, layer%n_rows, layer%n_cols, &
                         top_indices, top_scores, layer%top_k)

      ! Weighted sum
      total_score = sum(abs(top_scores))
      if (total_score < 1.0e-10_dp) total_score = 1.0_dp

      do k = 1, layer%top_k
        idx = top_indices(k)
        r = (idx - 1) / layer%n_cols + 1
        c = mod(idx - 1, layer%n_cols) + 1
        output_vec(:,q) = output_vec(:,q) + &
          (abs(top_scores(k)) / total_score) * layer%values(r, c, :)
      end do
    end do

    deallocate(scores, top_indices, top_scores)

  end subroutine query

  subroutine tucker_scores(layer, query_vec, scores, n_scores)
    type(MemoryLayer), intent(in) :: layer
    integer, intent(in) :: n_scores
    real(dp), intent(in)  :: query_vec(layer%hidden_dim)
    real(dp), intent(out) :: scores(n_scores)

    real(dp) :: q_rank(layer%n_ranks), c_rank(layer%n_ranks)
    real(dp) :: row_proj(layer%n_rows), col_proj(layer%n_cols)
    integer :: r, c, idx

    ! Project query to rank space: q_rank = row_keys * query
    q_rank = matmul(layer%row_keys, query_vec)

    ! Apply Tucker core
    c_rank = matmul(layer%tucker_core, q_rank)

    ! Row scores: row_proj(r) = dot(row_embed(:,r), q_rank)
    do r = 1, layer%n_rows
      row_proj(r) = dot_product(layer%row_embed(:, r), q_rank)
    end do

    ! Column scores: col_proj(c) = dot(col_embed(:,c), c_rank)
    do c = 1, layer%n_cols
      col_proj(c) = dot_product(layer%col_embed(:, c), c_rank)
    end do

    ! Assemble flattened scores as outer product
    idx = 1
    do r = 1, layer%n_rows
      do c = 1, layer%n_cols
        scores(idx) = row_proj(r) * col_proj(c)
        idx = idx + 1
      end do
    end do

  end subroutine tucker_scores

  subroutine top_k_indices(scores, indices, n_scores, k)
    integer, intent(in) :: n_scores, k
    real(dp), intent(in)  :: scores(n_scores)
    integer, intent(out)  :: indices(k)

    logical :: used(n_scores)
    integer :: i, j, best_j
    real(dp) :: best_score

    used = .false.

    do i = 1, k
      best_score = -huge(1.0_dp)
      best_j = 1
      do j = 1, n_scores
        if (.not. used(j) .and. scores(j) > best_score) then
          best_score = scores(j)
          best_j = j
        end if
      end do
      indices(i) = best_j
      used(best_j) = .true.
    end do

  end subroutine top_k_indices

  subroutine implicit_expand(values_in, linear_weights, values_out, n_v, n_h, n_expand)
    integer, intent(in) :: n_v, n_h, n_expand
    real(dp), intent(in)  :: values_in(n_v, n_h)
    real(dp), intent(in)  :: linear_weights(n_h, n_expand)
    real(dp), intent(out) :: values_out(n_v, n_h * n_expand)

    integer :: e

    ! Original values in first slot
    values_out(:, 1:n_h) = values_in

    ! Expanded: scaled linear projections
    do e = 1, n_expand - 1
      block
        real(dp) :: w_sum
        integer :: row_i
        w_sum = sum(linear_weights(:, e)) / real(n_h, dp)
        do row_i = 1, n_v
          values_out(row_i, e*n_h+1:(e+1)*n_h) = values_in(row_i, :) * w_sum
        end do
      end block
    end do

  end subroutine implicit_expand

  ! ---- Internal helpers ----

  subroutine compute_all_scores(layer, query_vec, scores)
    type(MemoryLayer), intent(in) :: layer
    real(dp), intent(in)  :: query_vec(layer%hidden_dim)
    real(dp), intent(out) :: scores(layer%n_rows, layer%n_cols)

    real(dp) :: q_rank(layer%n_ranks), c_rank(layer%n_ranks)
    real(dp) :: row_proj(layer%n_rows), col_proj(layer%n_cols)
    integer :: r, c

    ! Step 1: Project query to rank space via row_keys
    ! q_rank = row_keys * query  [n_ranks x 1]
    q_rank = matmul(layer%row_keys, query_vec)

    ! Step 2: Apply Tucker core decomposition
    ! c_rank = tucker_core * q_rank  [n_ranks x 1]
    c_rank = matmul(layer%tucker_core, q_rank)

    ! Step 3: Compute row scores via row embeddings
    do r = 1, layer%n_rows
      row_proj(r) = dot_product(layer%row_embed(:, r), q_rank)
    end do

    ! Step 4: Compute column scores via column embeddings
    do c = 1, layer%n_cols
      col_proj(c) = dot_product(layer%col_embed(:, c), c_rank)
    end do

    ! Step 5: Assemble score matrix via outer product
    do r = 1, layer%n_rows
      do c = 1, layer%n_cols
        scores(r, c) = row_proj(r) * col_proj(c)
      end do
    end do

  end subroutine compute_all_scores

  subroutine find_top_k_2d(scores, n_rows, n_cols, indices, top_scores, k)
    integer, intent(in) :: n_rows, n_cols, k
    real(dp), intent(in)  :: scores(n_rows, n_cols)
    integer, intent(out)  :: indices(k)
    real(dp), intent(out) :: top_scores(k)

    real(dp) :: flat(n_rows * n_cols)
    integer :: used(n_rows * n_cols)
    integer :: i, j, total, best_j
    real(dp) :: best_val

    total = n_rows * n_cols

    do i = 1, n_rows
      do j = 1, n_cols
        flat((i-1)*n_cols + j) = scores(i, j)
      end do
    end do

    used = 0

    do i = 1, k
      best_val = -huge(1.0_dp)
      best_j = 1
      do j = 1, total
        if (used(j) == 0 .and. flat(j) > best_val) then
          best_val = flat(j)
          best_j = j
        end if
      end do
      indices(i) = best_j
      top_scores(i) = best_val
      used(best_j) = 1
    end do

  end subroutine find_top_k_2d

end module sparse_memory

File: src/sparse_memory.rs
//! Safe Rust wrapper for the sparse_memory Fortran module.

/// Configuration for creating a new sparse memory layer.
#[derive(Debug, Clone)]
pub struct SparseMemoryConfig {
    pub n_rows: usize,
    pub n_cols: usize,
    pub hidden_dim: usize,
    pub n_ranks: usize,
    pub top_k: usize,
}

impl Default for SparseMemoryConfig {
    fn default() -> Self {
        Self {
            n_rows: 64,
            n_cols: 64,
            hidden_dim: 128,
            n_ranks: 16,
            top_k: 8,
        }
    }
}

/// Simple LCG random number generator (avoids rand dependency).
fn lcg_rand(seed: &mut u64) -> f64 {
    *seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1);
    ((*seed >> 33) as f64) / (1u64 << 31) as f64
}

/// A sparse memory layer with UltraMem-inspired architecture.
///
/// Pure Rust implementation that mirrors the Fortran sparse_memory module.
#[derive(Debug)]
pub struct SparseMemoryLayer {
    config: SparseMemoryConfig,
    values: Vec<Vec<Vec<f64>>>,     // (n_rows, n_cols, hidden_dim)
    row_keys: Vec<Vec<f64>>,        // (n_ranks, hidden_dim)
    row_embed: Vec<Vec<f64>>,       // (n_ranks, n_rows)
    col_keys: Vec<Vec<f64>>,        // (n_ranks, hidden_dim)
    col_embed: Vec<Vec<f64>>,       // (n_ranks, n_cols)
    tucker_core: Vec<Vec<f64>>,     // (n_ranks, n_ranks)
}

impl SparseMemoryLayer {
    /// Create a new sparse memory layer with pseudo-random initialization.
    pub fn new(config: SparseMemoryConfig) -> Self {
        let mut seed: u64 = 42;
        let scale = (2.0 / (config.n_ranks + config.hidden_dim) as f64).sqrt();

        let init_vec = |len: usize, s: &mut u64| -> Vec<f64> {
            (0..len).map(|_| (lcg_rand(s) - 0.5) * 2.0 * scale).collect()
        };

        let values = (0..config.n_rows)
            .map(|_| (0..config.n_cols)
                .map(|_| init_vec(config.hidden_dim, &mut seed))
                .collect())
            .collect();

        let row_keys = (0..config.n_ranks)
            .map(|_| init_vec(config.hidden_dim, &mut seed))
            .collect();

        let row_embed = (0..config.n_ranks)
            .map(|_| init_vec(config.n_rows, &mut seed))
            .collect();

        let col_keys = (0..config.n_ranks)
            .map(|_| init_vec(config.hidden_dim, &mut seed))
            .collect();

        let col_embed = (0..config.n_ranks)
            .map(|_| init_vec(config.n_cols, &mut seed))
            .collect();

        let tucker_core = (0..config.n_ranks)
            .map(|_| init_vec(config.n_ranks, &mut seed))
            .collect();

        Self { config, values, row_keys, row_embed, col_keys, col_embed, tucker_core }
    }

    /// Query the sparse memory layer.
    pub fn query(&self, query_vec: &[f64]) -> Vec<f64> {
        assert_eq!(query_vec.len(), self.config.hidden_dim);

        // Step 1: Project query to rank space via row_keys
        let q_rank: Vec<f64> = (0..self.config.n_ranks)
            .map(|r| self.row_keys[r].iter()
                .zip(query_vec.iter())
                .map(|(k, q)| k * q)
                .sum())
            .collect();

        // Step 2: Apply Tucker core
        let c_rank: Vec<f64> = (0..self.config.n_ranks)
            .map(|r| (0..self.config.n_ranks)
                .map(|k| self.tucker_core[r][k] * q_rank[k])
                .sum())
            .collect();

        // Step 3: Row scores
        let row_proj: Vec<f64> = (0..self.config.n_rows)
            .map(|r| self.row_embed.iter()
                .zip(q_rank.iter())
                .map(|(emb, qr)| emb[r] * qr)
                .sum())
            .collect();

        // Step 4: Column scores
        let col_proj: Vec<f64> = (0..self.config.n_cols)
            .map(|c| self.col_embed.iter()
                .zip(c_rank.iter())
                .map(|(emb, cr)| emb[c] * cr)
                .sum())
            .collect();

        // Step 5: Compute all scores and find top-k
        let mut scored: Vec<(usize, f64)> = Vec::new();
        for r in 0..self.config.n_rows {
            for c in 0..self.config.n_cols {
                let score = row_proj[r] * col_proj[c];
                let idx = r * self.config.n_cols + c;
                scored.push((idx, score));
            }
        }

        scored.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        scored.truncate(self.config.top_k);

        let total_score: f64 = scored.iter().map(|(_, s)| s.abs()).sum::<f64>().max(1e-10);
        let mut output = vec![0.0; self.config.hidden_dim];
        for (idx, score) in &scored {
            let r = idx / self.config.n_cols;
            let c = idx % self.config.n_cols;
            let w = score.abs() / total_score;
            for d in 0..self.config.hidden_dim {
                output[d] += w * self.values[r][c][d];
            }
        }

        output
    }

    /// Get the configuration.
    pub fn config(&self) -> &SparseMemoryConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_layer() {
        let config = SparseMemoryConfig {
            n_rows: 4, n_cols: 4, hidden_dim: 8,
            n_ranks: 2, top_k: 2,
        };
        let layer = SparseMemoryLayer::new(config);
        assert_eq!(layer.config().n_rows, 4);
    }

    #[test]
    fn test_query() {
        let config = SparseMemoryConfig {
            n_rows: 4, n_cols: 4, hidden_dim: 8,
            n_ranks: 2, top_k: 2,
        };
        let layer = SparseMemoryLayer::new(config);
        let query = vec![1.0, 0.5, -0.3, 0.7, -0.2, 0.4, 0.1, -0.6];
        let output = layer.query(&query);
        assert_eq!(output.len(), 8);
        let norm: f64 = output.iter().map(|x| x * x).sum::<f64>().sqrt();
        assert!(norm > 0.0, "Query output should be non-zero");
    }
}
